### PR TITLE
`tdesktop-beta`: init at 3.6.0 & `tdesktop`: fix executable path for shortcuts

### PR DESCRIPTION
**Pull request by psydvl**
###### Description of changes
`tdesktop`: fix executable path for shortcuts
Which allows autogenerated .desktop files for mime defaults keeping right path
Also in `tdesktop` fixed some small things
`tdesktop-beta`: init at 3.6.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by yshym — COMMENTED** [review]
Have you considered separating into 3 files: `tdesktop/generic.nix`, `tdesktop/default.nix`, `tdesktop/beta.nix`? Won't it grow into too much `if` statements with future releases if not separated?

File: pkgs/applications/networking/instant-messengers/telegram/tdesktop/update.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def update_file(relpath, version, sha256, rev=None):
         for line in f:
             result = line
             result = re.sub(r'^  version = ".+";', f'  version = "{version}";', result)
-            result = re.sub(r'^    sha256 = ".+";', f'    sha256 = "{sha256}";', result)
+            result = re.sub(r'^(\s*)sha256 = ".+";', f'\g<1>sha256 = "{sha256}";', result)
```
Comment (yshym) [inline pkgs/applications/networking/instant-messengers/telegram/tdesktop/update.py:58]:
```suggestion
            result = re.sub(r'^(\s*)sha256 = ".+";', fr'\g<1>sha256 = "{sha256}";', result)
```

File: pkgs/applications/networking/instant-messengers/telegram/tdesktop/update.py [hunk 2]
```diff
@@ -66,8 +66,11 @@ def update_file(relpath, version, sha256, rev=None):
     tdesktop_version = tdesktop_tag.lstrip('v')
     tdesktop_hash = nix_prefetch_git('https://github.com/telegramdesktop/tdesktop.git', tdesktop_tag)
     update_file('default.nix', tdesktop_version, tdesktop_hash)
+    tdesktop_beta_tag = github_api_request('repos/telegramdesktop/tdesktop/releases')[0]['tag_name']
```
[the remark this answers is not in the document]
Comment (yshym) [reply pkgs/applications/networking/instant-messengers/telegram/tdesktop/update.py:69]: Ok, I can see, that `/releases/latest` is about the latest non-prerelease tag. As far as I understand, the point of `-beta` package is just to allow users to get the bleeding edge version of the app. I thought if there are any use cases when we wouldn't want a beta package to update to a stable version, but can't see ones at least for this package

**Review by yshym — COMMENTED** [review]
Commit messages should contain information that the stable package was updated to 3.6.0

File: pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioOutputALSA.cpp \
+      --replace '"libasound.so.2"' '"${alsa-lib}/lib/libasound.so.2"'
+    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioPulse.cpp \
+      --replace '"libpulse.so.0"' '"${libpulseaudio}/lib/libpulse.so.0"'
+    substituteInPlace Telegram/lib_webview/webview/platform/linux/webview_linux_webkit_gtk.cpp \
+      --replace '"libwebkit2gtk-4.0.so.37"' '"${webkitgtk}/lib/libwebkit2gtk-4.0.so.37"'
+    substituteInPlace Telegram/SourceFiles/core/application.cpp \
+      --replace "cExeDir() + cExeName()" '"telegram-desktop"'
+    substituteInPlace Telegram/SourceFiles/platform/linux/specific_linux.cpp \
+      --replace "cExeDir() + cExeName()" '"telegram-desktop"'
```
Comment (yshym) [inline pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix:100]: Can you please explain in more detail, what exactly do these lines fix?

Comment (FliegendeWurst) [reply pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix:100]: I think these substitutions are supposed to modify the autostart .desktop file.
Currently, the autostart file will contain a store path:
```Exec=/nix/store/jqji1v34cl1rpayfjd9i0c8db8rcpys6-telegram-desktop-3.6.0/bin/.telegram-desktop-wrapped -workdir /home/arne/.local/share/TelegramDesktop/ -autostart```
After substituting, it should contain `Exec=telegram-desktop -workdir ....`

File: pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    libdatrie
+    libsysprof-capture
+    libpsl
+    brotli
+    microsoft_gsl
+    rlottie
+  ];
+
+  cmakeFlags = [
+    "-Ddisable_autoupdate=ON"
```
Comment (FliegendeWurst) [inline pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix:159]:
```suggestion
    "-DTDESKTOP_DISABLE_AUTOUPDATE=ON"
```

**Comment by nomeata** [comment]
This says init at 3.6.1, but 4.3.1 (non-beta) is already in `unstable`. Can this be closed?

**Pull request closed without merging**